### PR TITLE
Add classification API support for VB members, locals and parameters, and tweak C# implementation

### PR DESCRIPTION
**Pull request by DustinCampbell**
This is a follow up to https://github.com/dotnet/roslyn/pull/24931 that tweaks C# classification and adds support for Visual Basic. This only affects the classification APIs and does is not (yet) exposed in Visual Studio. This is related to https://github.com/dotnet/roslyn/issues/3976.

The design tweaks I've made to the original C# implementation (#24931) are:

* Extension methods are only classified at declarations or when used as a reduced form. If an extension method is called as a static method invocation (i.e. `Enumerable.Select(...)`), it will be classified as an ordinary method.
* Local constants are now classified as constants rather than locals.
* `EnumFieldName` has been renamed to `EnumMemberName` to better match the rest of the Roslyn API (i.e. `EnumMemberDeclarationSyntax`).

Note that extension method declarations in VB are impossible to classify syntactically. So, they VB methods are classified as methods and the semantic classifier adds an additional classification for extension methods.

In addition, this PR unifies the classification unit test infrastructure quite a bit. It might be useful to review one commit at a time if it's too cumbersome.

cc @mkrueger

### CODE REVIEW

File: src/EditorFeatures/TestUtilities/Classification/AbstractClassifierTests.cs [hunk 1]
```diff
@@ -1,176 +1,298 @@
...
-            return ClassificationBuilder.Identifier(value);
-        }
+        protected static ParseOptions[] ParseOptions(params ParseOptions[] options) => options;
 
         [DebuggerStepThrough]
-        protected Tuple<string, string> Inactive(string value)
-        {
-            return ClassificationBuilder.Inactive(value);
-        }
+        protected FormattedClassification Struct(string text) => ClassificationBuilder.Struct(text);
```
Comment (Pilchie) [inline src/EditorFeatures/TestUtilities/Classification/AbstractClassifierTests.cs]: Would a `using static` of `ClassificationBuilder` avoid the need to have all of these?

**Comment by DustinCampbell** [comment]
@dotnet-bot retest this please

File: src/EditorFeatures/TestUtilities/Classification/AbstractClassifierTests.cs (hunk 1 above)
Comment (DustinCampbell) [reply src/EditorFeatures/TestUtilities/Classification/AbstractClassifierTests.cs]: heh-heh. Yes, I think you're right. I'll take a look at that.

File: src/Workspaces/Core/Portable/Classification/ClassificationTypeNames.cs [hunk 2]
```diff
@@ -28,7 +28,7 @@ public static class ClassificationTypeNames
         public const string TypeParameterName = "type parameter name";
 
         public const string FieldName = "field name";
-        public const string EnumFieldName = "enum field name";
+        public const string EnumMemberName = "enum member name";
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/Classification/ClassificationTypeNames.cs]: Do we have to update EditorColors.xml in the VS repo for theming?

Comment (dpoeschl) [reply src/Workspaces/Core/Portable/Classification/ClassificationTypeNames.cs]: I'm not sure if the right thing happened there when the C# ones were introduced either.

Comment (DustinCampbell) [reply src/Workspaces/Core/Portable/Classification/ClassificationTypeNames.cs]: We don't need to update EditorColors.xml yet. This is just an API-only change. In VS, these all map back to "identifier" for now. When we introduce this feature for VS we'll update it for theming.

**Comment by jinujoseph** [comment]
@Pilchie for ask mode approval for 15.7.P2

**Comment by Pilchie** [comment]
Approved for 15.7 P2

**Comment by DustinCampbell** [comment]
Thanks all!

**Pull request merged**